internal/db/db_certificates.go

```diff
@@ -0,0 +1,293 @@
+	addCertificateToCSRsStmt = "UPDATE %s SET certificate_id=$Certificate.certificate_id, status=$CertificateRequest.status WHERE id==$CertificateRequest.id or csr==$CertificateRequest.csr"
+	getCertificateStmt       = "SELECT &Certificate.* FROM %s WHERE certificate_id==$Certificate.certificate_id or certificate==$Certificate.certificate"
+	listCertificatesStmt     = "SELECT &Certificate.* FROM %s"
+	deleteCertificateStmt    = "DELETE FROM %s WHERE certificate_id=$Certificate.certificate_id or certificate=$Certificate.certificate"
+
+	getCertificateChainStmt = `WITH RECURSIVE cert_chain AS (
+	-- Initial query: Start search from the end certificate
+    SELECT certificate_id, certificate, issuer_id
+    FROM %s
+    WHERE certificate_id = $Certificate.certificate_id or certificate = $Certificate.certificate
+    
+    UNION ALL
+    
+    -- Recursive Query: Move up the chain until issuer_id is 0 (root)
+    SELECT certs.certificate_id, certs.certificate, certs.issuer_id
+    FROM certificates certs
+    JOIN cert_chain
+      ON certs.certificate_id = cert_chain.issuer_id
+)
+SELECT * FROM cert_chain;`
+)
+
+// ListCertificateRequests gets every CertificateRequest entry in the table.
+func (db *Database) ListCertificates() ([]Certificate, error) {
+	stmt, err := sqlair.Prepare(fmt.Sprintf(listCertificatesStmt, db.certificatesTable), Certificate{})
+	if err != nil {
+		return nil, err
+	}
+	var certificates []Certificate
+	err = db.conn.Query(context.Background(), stmt).GetAll(&certificates)
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return certificates, nil
+		}
+		return nil, err
+	}
+	return certificates, nil
+}
+
+// GetCertificateByID gets a certificate row from the repository from a given ID.
+func (db *Database) GetCertificateByID(id int) (*Certificate, error) {
+	certRow := Certificate{
+		CertificateID: id,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCertificateStmt, db.certificatesTable), Certificate{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, certRow).Get(&certRow)
+	if err != nil {
+		return nil, err
+	}
+	return &certRow, nil
+}
+
+// GetCertificateByID gets a certificate row from the repository from a given ID.
+func (db *Database) GetCertificateByCertificatePEM(cert string) (*Certificate, error) {
+	certRow := Certificate{
+		CertificatePEM: cert,
+	}
+	stmt, err := sqlair.Prepare(fmt.Sprintf(getCertificateStmt, db.certificatesTable), Certificate{})
+	if err != nil {
+		return nil, err
+	}
+	err = db.conn.Query(context.Background(), stmt, certRow).Get(&certRow)
+	if err != nil {
+		return nil, err
+	}
+	return &certRow, nil
+}
+
+// AddCertificateChainToCertificateRequestByCSR adds a new certificate chain to a row for a given CSR string.
+func (db *Database) AddCertificateChainToCertificateRequestByCSR(csrPEM string, certPEM string) error {
```

Comment:
This function and the next one do almost completely the same thing, could we reuse most of the code by extracting a function and passing the CSR after looking it up either by ID or CSR?